Python 3.9 migration

sherwinlu opened the pull request:
Migrating the project to be Python 3.9 compatible.

sherwinlu committed:
initial migration to python 3.9 and upgrading 3rd party library dependencies to the latest

sherwinlu committed:
changing the repo for circleci

sherwinlu committed:
fixing an issue with the upgrade from Python 2 to 3

sherwinlu committed:
fixing an issue with the build file

sherwinlu committed:
removing the ROOT_URLCONF

sherwinlu committed:
removed the ROOT_URLCONF

sherwinlu committed:
rewriting the code to use Python 3 syntax

sherwinlu committed:
updating the boto3 library

gurkansokat commented on .circleci/config.yml:
```diff
@@ -3,24 +3,22 @@ workflows:
+       - test-python-3-9
 jobs:
-  test-python-3-7: &test-template
+  test-python-3-9: &test-template
```
I think we can remove the test-template.

gurkansokat commented on .circleci/config.yml:
```diff
@@ -3,24 +3,22 @@ workflows:
-    docker:
-      - image: circleci/python:2.7.16
+            . venv/bin/activate
+            python -m django test --settings=eb_sqs.test_settings
```
Nit. 
New line at the end of file.

gurkansokat commented on eb_sqs/auto_tasks/service.py:
```diff
@@ -54,12 +55,12 @@ def _auto_task_wrapper(module_name, class_name, func_name, *args, **kwargs):
             else:
                 # by default log an error
-                logger.error('Reached max retries in auto task {}.{}.{} with error: {}'.format(module_name, class_name, func_name, repr(exc)))
+                logger.error('Reached max retries in auto task {}.{}.{} with error: {}'.format(module_name, class_name,
```
Should we convert these to f-string way now?

gurkansokat replied:
There are other places that we use the old .format way, but not sure if it is too much work or not to convert all to f-string. What do you think?

sherwinlu replied:
i think we can just leave it as is. there's too many lines of code that use the old style and it's not worth the effort to convert imo - unless Python plans to not support that old style format anytime soon.

gurkansokat commented on eb_sqs/worker/worker_task.py:
```diff
@@ -76,13 +74,11 @@ def copy(self, use_serialization):
     @staticmethod
-    def deserialize(msg):
-        # type: (unicode) -> WorkerTask
+    def deserialize(msg: str):
```
Need to add -> WorkerTask for return?

gurkansokat commented on eb_sqs/worker/worker_task.py:
```diff
@@ -30,15 +30,13 @@ def __init__(self, id, group_id, queue, func, args, kwargs, max_retries, retry,
 
-    def serialize(self):
-        # type: () -> unicode
+    def serialize(self) -> str:
```
Not related to this particular line, but below we have missed the syntax update for the "copy" function

gurkansokat commented on eb_sqs/worker/worker.py:
```diff
@@ -69,19 +68,19 @@ def execute(self, msg):
-    def delay(self, group_id, queue_name, func, args, kwargs, max_retries, use_pickle, delay, execute_inline):
-        # type: (unicode, unicode, Any, tuple, dict, int, bool, int, bool) -> Any
-        worker_task = WorkerTask(str(uuid.uuid4()), group_id, queue_name, func, args, kwargs, max_retries, 0, None, use_pickle)
+    def delay(self, group_id: str, queue_name: str, func, args: tuple, kwargs: dict, max_retries: int, use_pickle: bool,
```
func parameter missing type. (func: Any)

gurkansokat commented on eb_sqs/worker/service.py:
```diff
@@ -4,8 +4,13 @@
+from typing import Any
 
 import boto3
+from boto.sqs.message import Message
```
are these the right from packages? My PyCharm was complaining and I couldn't find an example of this anywhere.

sherwinlu replied:
yeah i'm getting the same errors too, but when i look at the documentation, it looks correct - https://boto3.amazonaws.com/v1/documentation/api/latest/guide/sqs.html#sqs

sherwinlu replied:
actually when i looked again boto = boto 2.0 and we're using boto 3.0 so the reference is incorrect. however, i can't find Queue nor Message in the boto3 (and botocore) packages. So I've left it out for now.

gurkansokat commented on eb_sqs/auto_tasks/service.py:
```diff
@@ -54,12 +55,12 @@ def _auto_task_wrapper(module_name, class_name, func_name, *args, **kwargs):
 class AutoTaskService(object):
-    def register_task(self, method, queue_name=None, max_retries=None):
-        # type: (Any, str, int) -> None
+    def register_task(self, method, queue_name: str = None, max_retries: int = None):
```
"method -> Any" or we skip it on purpose?

gurkansokat commented on eb_sqs/auto_tasks/exceptions.py:
```diff
@@ -1,12 +1,10 @@
 class RetryableTaskException(Exception):
-    def __init__(self, inner, delay=None, count_retries=None, max_retries_func=None):
-        # type: (Exception, int, bool, Any) -> None
+    def __init__(self, inner: Exception, delay: int = None, count_retries: bool = None, max_retries_func=None):
```
max_retries_func: Any=None?

gurkansokat commented:
It seems like we are still expecting to pass circleci tests for python 2.7 and 3.7 still. Are those in project settings?

alexeyts commented on eb_sqs/auto_tasks/service.py:
```diff
@@ -24,7 +24,8 @@ def _auto_task_wrapper(module_name, class_name, func_name, *args, **kwargs):
         auto_task_executor_service = _AutoTaskExecutorService(func_name)
-        instance = class_(auto_task_service=auto_task_executor_service)  # instantiate class using _AutoTaskExecutorService
+        instance = class_(
+            auto_task_service=auto_task_executor_service)  # instantiate class using _AutoTaskExecutorService
```
nit: indentation - closing bracket and comment could be on next line (and comment can probably be removed altogether)

just wanted to add a comment on this PR! 😁

gurkansokat replied:
I thought GitHub was messing with me when I saw the name.. 😁 Good to see you watching...

sherwinlu commented on eb_sqs/auto_tasks/service.py:
```diff
@@ -82,14 +83,12 @@ def _auto_task_wrapper_invoker(*args, **kwargs):
 
-    def register_task(self, method, queue_name=None, max_retries=None):
-        # type: (Any, str, int) -> None
+    def register_task(self, method, queue_name: str = None, max_retries: int = None):
```
nope, was an oversight.

sherwinlu committed:
updating based on PR comments

sherwinlu commented:
@gurkansokat not sure why it's still looking for the python 2-7 and 3-7 tests. I removed them from the config.yml for circleci and replaced it with the python-3-9 test

sherwinlu committed:
fixing an error

sherwinlu committed:
fixing an error

sherwinlu committed:
bumping the boto3 library

sherwinlu committed:
removing the reference to boto as it's from boto2

sherwinlu commented on eb_sqs/worker/worker_factory.py:
```diff
@@ -10,17 +10,14 @@ class WorkerFactory(object):
 
     @staticmethod
     def default():
-        # type: () -> WorkerFactory
```
It doesn't like it when I put WorkerFactory as the return type. It complains about not being able to find that class.

sherwinlu committed:
rewriting the metaclass format

sherwinlu committed:
removing python2 imports

sherwinlu committed:
testing the build

sherwinlu committed:
removing the common-utils
